### PR TITLE
Update Virtlet and CRI proxy for k8s 1.6

### PR DESCRIPTION
**Pull request by ivan4th**
- [x] Make it possible to start Virtlet on k8s 1.6 in semi-manual mode (1)
- [x] Add hack for `\"enforceNodeAllocatable\":[]` in case of DIND (2) 
- [x] Fix `nodeAffinity` - it's being ignored in daemonset definition
- [x] Add RBAC defs to yamls (non-permissive version)
- [x] Fix DNS in Virtlet pod
- [x] Fix kubelet restart during criproxy bootstrap (3)
- [x] Add test for ImageRef in PullImage response (4)
- [x] Add test for Metadata.Name in ContainerStatus response (5)
- [x] Fix flexvolume
- [x] Fix tests
- [x] Fix CI

Details

1. After initial impl the following extra steps need to be done, besides ones described in `running-local-environment.md`:
```
# after bringing up the cluster, before applying Virtlet yaml
# set up RBAC
kubectl create clusterrolebinding permissive-binding \
   --clusterrole=cluster-admin \
   --user=admin \
   --user=kubelet \
   --group=system:serviceaccounts
# delete kube-node-2 because of nodeAffinity problems
kubectl delete node kube-node-2
```
After creating Virtlet daemonset, enter `kube-node-1` container and kill `sleep Infinity` there. After starting imager-server/image-service, take service IP from `kubectl describe -n kube-system image-service` and put it instead of `image-service.kube-system` in cirros-vm.yaml.

2. There's bug in kubelet 1.6, it doesn't serialize empty `enforceNodeAllocatable` array currently employed in DIND cgroup workaround for 1.6 when responding to `/configz` request. Currently setting empty `enforceNodeAllocatable` is hardcoded in CRI proxy bootstrap but it should only be done for DIND.

3. As of now kubelet defaults to CRI, so CRI proxy bootstrap container doesn't die during kubelet restart, so it keeps sitting in `sleep Infinity`.  It should detect kubelet restart properly instead.

4. That's an API change in k8s 1.6, the code is already fixed but tests are not updated yet. This should be tested for both criproxy and Virtlet itself.

5. This part of API didn't really change but extra validation was added on kubelet side, so Virtlet started failing checks during `CreateContainer`. The code is already fixed but need to update tests.

---
This change is [Reviewable](https://reviewable.io/reviews/mirantis/virtlet/261)

### CODE REVIEW

**Comment by jellonek** [comment]
Reviewed 44 of 44 files at r1.
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[pkg/criproxy/bootstrap.go, line 43 at r1](https://reviewable.io:443/reviews/mirantis/virtlet/261#-Ki9NUbNW9V7fNeDb1M5:-Ki9NUbNW9V7fNeDb1M6:b-7su253) ([raw file](https://github.com/mirantis/virtlet/blob/8e2732adb97ea88b184d6c815cd72c541edabb4c/pkg/criproxy/bootstrap.go#L43)):*
> ```Go
> 
> 	// FIXME: use client-go
> 	// In particular, see https://github.com/kubernetes/client-go/blob/master/examples/in-cluster/main.go
> ```

This comment seems to be outdated ;)

---

*[pkg/criproxy/bootstrap.go, line 169 at r1](https://reviewable.io:443/reviews/mirantis/virtlet/261#-Ki9Nk_rfgvGR90AQVVM:-Ki9Nk_rfgvGR90AQVVN:blweklj) ([raw file](https://github.com/mirantis/virtlet/blob/8e2732adb97ea88b184d6c815cd72c541edabb4c/pkg/criproxy/bootstrap.go#L169)):*
> ```Go
> 	kubeletCfg, err := b.retrieveKubeletConfig()
> 	if err != nil {
> 		glog.V(2).Infof("can't retrieve kubelet config yet: %v", err)
> ```

Message should start from capital letter.

---

*[pkg/criproxy/proxy.go, line 329 at r1](https://reviewable.io:443/reviews/mirantis/virtlet/261#-Ki9OoKEaC0p5LBK3g-M:-Ki9OoKEaC0p5LBK3g-N:b-asfmme) ([raw file](https://github.com/mirantis/virtlet/blob/8e2732adb97ea88b184d6c815cd72c541edabb4c/pkg/criproxy/proxy.go#L329)):*
> ```Go
> 	}
> 	// TODO: check if the server is present
> 	r.server.GracefulStop()
> ```

If now it's graceful - is above comment still correct?

---

*[pkg/criproxy/proxy_test.go, line 1145 at r1](https://reviewable.io:443/reviews/mirantis/virtlet/261#-Ki9PbPko_Tv0gJl0oAL:-Ki9PbPko_Tv0gJl0oAM:bzg1e0o) ([raw file](https://github.com/mirantis/virtlet/blob/8e2732adb97ea88b184d6c815cd72c541edabb4c/pkg/criproxy/proxy_test.go#L1145)):*
> ```Go
> func TestCriProxyNoStartupRace(t *testing.T) {
> 	if inTravis() {
> 		t.Skip("apparently there's still a startup race in CRI proxy, but it only manifests itself on slower machines")
> ```

If it's race, can we wait for criproxy longer, instead of skipping these tests?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mirantis/virtlet/261)*

**Comment by ivan4th** [comment]
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[pkg/criproxy/bootstrap.go, line 43 at r1](https://reviewable.io:443/reviews/mirantis/virtlet/261#-Ki9NUbNW9V7fNeDb1M5:-KiANz9DmnAayfE4L-yU:bkhcd3h) ([raw file](https://github.com/mirantis/virtlet/blob/8e2732adb97ea88b184d6c815cd72c541edabb4c/pkg/criproxy/bootstrap.go#L43)):*
<details><summary><i>Previously, jellonek (Piotr Skamruk) wrote…</i></summary><blockquote>

This comment seems to be outdated ;)
</blockquote></details>

Will remove

---

*[pkg/criproxy/bootstrap.go, line 169 at r1](https://reviewable.io:443/reviews/mirantis/virtlet/261#-Ki9Nk_rfgvGR90AQVVM:-KiAO2VZMFIKt7GAI4Pm:bhy9q7g) ([raw file](https://github.com/mirantis/virtlet/blob/8e2732adb97ea88b184d6c815cd72c541edabb4c/pkg/criproxy/bootstrap.go#L169)):*
<details><summary><i>Previously, jellonek (Piotr Skamruk) wrote…</i></summary><blockquote>

Message should start from capital letter.
</blockquote></details>

Will fix

---

*[pkg/criproxy/proxy.go, line 329 at r1](https://reviewable.io:443/reviews/mirantis/virtlet/261#-Ki9OoKEaC0p5LBK3g-M:-KiAO4Q3VDMPk2-em0iu:b97cpr1) ([raw file](https://github.com/mirantis/virtlet/blob/8e2732adb97ea88b184d6c815cd72c541edabb4c/pkg/criproxy/proxy.go#L329)):*
<details><summary><i>Previously, jellonek (Piotr Skamruk) wrote…</i></summary><blockquote>

If now it's graceful - is above comment still correct?
</blockquote></details>

Will remove the comment

---

*[pkg/criproxy/proxy_test.go, line 1145 at r1](https://reviewable.io:443/reviews/mirantis/virtlet/261#-Ki9PbPko_Tv0gJl0oAL:-KiAO6CPPe1H0VhDXWUu:b-wdoj4k) ([raw file](https://github.com/mirantis/virtlet/blob/8e2732adb97ea88b184d6c815cd72c541edabb4c/pkg/criproxy/proxy_test.go#L1145)):*
<details><summary><i>Previously, jellonek (Piotr Skamruk) wrote…</i></summary><blockquote>

If it's race, can we wait for criproxy longer, instead of skipping these tests?
</blockquote></details>

No, it's more complex than that. Waiting doesn't help, the problem is that `ListContainers()` should never fail when 2nd CRI impl is not connected yet, but under some circumstances it still may fail when it's being connected to. We'll need to fix this at some point -- will need to make an issue for this.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mirantis/virtlet/261)*

**Comment by ivan4th** [comment]
Review status: 41 of 44 files reviewed at latest revision, 4 unresolved discussions.

---

*[pkg/criproxy/proxy.go, line 329 at r1](https://reviewable.io:443/reviews/mirantis/virtlet/261#-Ki9OoKEaC0p5LBK3g-M:-KiAOufOlfIyGskAk8me:b-8wuyri) ([raw file](https://github.com/mirantis/virtlet/blob/8e2732adb97ea88b184d6c815cd72c541edabb4c/pkg/criproxy/proxy.go#L329)):*
<details><summary><i>Previously, ivan4th (Ivan Shvedunov) wrote…</i></summary><blockquote>

Will remove the comment
</blockquote></details>

Oops, nope, that's about different problem. Need to check whether the server was started properly (not critical)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mirantis/virtlet/261)*

**Comment by jellonek** [comment]
:lgtm:

---

Reviewed 5 of 5 files at r2.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/mirantis/virtlet/261#-:-KiASROUZ0qczm3p0YiS:bnfp4nl)*

**Comment by vefimova** [comment]
Great Job! Only two comments.

---

Reviewed 39 of 44 files at r1, 5 of 5 files at r2.
Review status: all files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[build/cmd.sh, line 164 at r2](https://reviewable.io:443/reviews/mirantis/virtlet/261#-KiB2uGvSvoWnZulh_pn:-KiB2uGvSvoWnZulh_po:b-reeegb) ([raw file](https://github.com/mirantis/virtlet/blob/dba566b028b98d2d29726cd02f9d5751cf7641ee/build/cmd.sh#L164)):*
> ```Shell
>         vcmd "$*"
>         ;;
>     vsh)
> ```

Would be good to add few nodes regarding this new cmd to doc.

---

*[pkg/flexvolume/flexvolume.go, line 155 at r2](https://reviewable.io:443/reviews/mirantis/virtlet/261#-KiB37Pn6pB-WCocyqMx:-KiB37Pn6pB-WCocyqMy:b620g8s) ([raw file](https://github.com/mirantis/virtlet/blob/dba566b028b98d2d29726cd02f9d5751cf7641ee/pkg/flexvolume/flexvolume.go#L155)):*
> ```Go
> }
> 
> // Invocation: <driver executable> getvolumename <json options>
> ```

As I see we don't need "getVolumeName" function as it's only used inside default attacher interface of flex-volume plugin,
which we intentionally override with empty ones in our driver and return success:

https://github.com/kubernetes/kubernetes/blob/bf984aa328f3d3e5f8956f7d8e65c13c52102426/pkg/volume/flexvolume/attacher-defaults.go#L49

So basically it even isn't going to be called, if I'm not missing smth. Also it looks strange to return some artificial volume name on request of kubelet without tracking further actions.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mirantis/virtlet/261#-:-KiB5OV4RNy8JPExv8zW:bnxz4vq)*

**Comment by ivan4th** [comment]
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[build/cmd.sh, line 164 at r2](https://reviewable.io:443/reviews/mirantis/virtlet/261#-KiB2uGvSvoWnZulh_pn:-KiBTgVxy3d6LJ20f62a:bef2vhg) ([raw file](https://github.com/mirantis/virtlet/blob/dba566b028b98d2d29726cd02f9d5751cf7641ee/build/cmd.sh#L164)):*
<details><summary><i>Previously, vefimova (Victoria Efimova) wrote…</i></summary><blockquote>

Would be good to add few nodes regarding this new cmd to doc.
</blockquote></details>

ok will do this

---

*[pkg/flexvolume/flexvolume.go, line 155 at r2](https://reviewable.io:443/reviews/mirantis/virtlet/261#-KiB37Pn6pB-WCocyqMx:-KiBTiD9yroa1OU2y60e:bo5lgz4) ([raw file](https://github.com/mirantis/virtlet/blob/dba566b028b98d2d29726cd02f9d5751cf7641ee/pkg/flexvolume/flexvolume.go#L155)):*
<details><summary><i>Previously, vefimova (Victoria Efimova) wrote…</i></summary><blockquote>

As I see we don't need "getVolumeName" function as it's only used inside default attacher interface of flex-volume plugin,
which we intentionally override with empty ones in our driver and return success:

https://github.com/kubernetes/kubernetes/blob/bf984aa328f3d3e5f8956f7d8e65c13c52102426/pkg/volume/flexvolume/attacher-defaults.go#L49

So basically it even isn't going to be called, if I'm not missing smth. Also it looks strange to return some artificial volume name on request of kubelet without tracking further actions.
</blockquote></details>

actually getvolume gets called (posted a log to chat)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mirantis/virtlet/261)*

**Comment by vefimova** [comment]
Review status: 43 of 45 files reviewed at latest revision, 2 unresolved discussions.

---

*[pkg/flexvolume/flexvolume.go, line 155 at r2](https://reviewable.io:443/reviews/mirantis/virtlet/261#-KiB37Pn6pB-WCocyqMx:-KiCAhkf95G55TF_q2X8:b-nkf0hr) ([raw file](https://github.com/mirantis/virtlet/blob/dba566b028b98d2d29726cd02f9d5751cf7641ee/pkg/flexvolume/flexvolume.go#L155)):*
<details><summary><i>Previously, ivan4th (Ivan Shvedunov) wrote…</i></summary><blockquote>

actually getvolume gets called (posted a log to chat)
</blockquote></details>

aha, thought that "mountdevice" wouldn't be called if we return devicePath="", but it is also called  and inside it we make call to getvolumename, so as we returned "Not supported",  we don't expect here flexvolume plugin to perform any actions (https://github.com/kubernetes/kubernetes/blob/bf984aa328f3d3e5f8956f7d8e65c13c52102426/pkg/volume/flexvolume/attacher.go#L91), so it still doesn't matter what we return and default func can be used safely.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mirantis/virtlet/261)*

**Comment by vefimova** [comment]
Reviewed 2 of 2 files at r3.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/mirantis/virtlet/261)*

**Comment by ivan4th** [comment]
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[pkg/flexvolume/flexvolume.go, line 155 at r2](https://reviewable.io:443/reviews/mirantis/virtlet/261#-KiB37Pn6pB-WCocyqMx:-KiEb35M0w6pQdfsj6dy:b223npm) ([raw file](https://github.com/mirantis/virtlet/blob/dba566b028b98d2d29726cd02f9d5751cf7641ee/pkg/flexvolume/flexvolume.go#L155)):*
<details><summary><i>Previously, vefimova (Victoria Efimova) wrote…</i></summary><blockquote>

aha, thought that "mountdevice" wouldn't be called if we return devicePath="", but it is also called  and inside it we make call to getvolumename, so as we returned "Not supported",  we don't expect here flexvolume plugin to perform any actions (https://github.com/kubernetes/kubernetes/blob/bf984aa328f3d3e5f8956f7d8e65c13c52102426/pkg/volume/flexvolume/attacher.go#L91), so it still doesn't matter what we return and default func can be used safely.
</blockquote></details>

default `getvolumename` can be used safely indeed, but implementing it produces more informative names under `/var/lib/kubelet/pods/...`. Moreover, flexvolume examples do implement it :) Anyway, we may discuss it later and correct in followup PRs if necessary I think.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mirantis/virtlet/261)*

**Comment by ivan4th** [comment]
Merging, because postponing this PR and basing further work on master could lead to merge conflicts.

**Comment by miha-plesko** [comment]
@ivan4th is it safe for me to switch to k8t 1.6 in my development environment or should I stay on 1.5 little longer?

**Comment by ivan4th** [comment]
@miha-plesko well, I'd say it's much better to switch to 1.6 at this point, as `mirantis/virtlet` image is for 1.6 now.

**Comment by jellonek** [comment]
@miha-plesko it's on master now, because we tested it, several times manually as on CI (travis).
Also it has a lot of cleanup in code (especially in pointers usage).

**Comment by miha-plesko** [comment]
Yup, I've seen the commits. I also appreciate your decision to take `demo.sh` under git control. The reason I asked you about current status is since I wasn't sure if the "old" yaml format for init containers that is used in `examples` folder  would still work on 1.6. I'm migrating to 1.6 now, thanks for the reply.
